util/buildinfo/buildinfo.go

```diff
@@ -0,0 +1,135 @@
+package buildinfo
+
+import (
+	"context"
+	"encoding/json"
+	"sort"
+
+	"github.com/docker/distribution/reference"
+	"github.com/moby/buildkit/exporter/containerimage/exptypes"
+	"github.com/moby/buildkit/source"
+	"github.com/moby/buildkit/util/urlutil"
+	"github.com/pkg/errors"
+)
+
+// Merge combines and fixes build info from image config
+// key moby.buildkit.buildinfo.v0.
+func Merge(ctx context.Context, buildInfo map[string]string, imageConfig []byte) ([]byte, error) {
+	icbi, err := imageConfigBuildInfo(imageConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// Iterate and combine build sources
+	mbis := map[string]exptypes.BuildInfo{}
+	for srcs, di := range buildInfo {
+		src, err := source.FromString(srcs)
+		if err != nil {
+			return nil, err
+		}
+		switch sid := src.(type) {
+		case *source.ImageIdentifier:
+			for _, bi := range icbi {
+				// Use original user input from image config
+				if bi.Type == exptypes.BuildInfoTypeDockerImage && bi.Alias == sid.Reference.String() {
```

Comment:
this reference does not need `TagNameOnly` ?

Reply:
no we want to match the alias which is the actual ref.

Reply:
What is "actual ref" in here. `TagNameOnly` would mean we normalize both sides.

Reply:
The pseudo-ref from alias sorry which is retrieved from `ResolveImageConfig` in the frontend and will match the one from LLB.